os/fs/driver/mtd/smart.c

```diff
@@ -5068,51 +5117,55 @@ static int smart_journal_checkout_process(FAR struct smart_struct_s *dev, journa
 static int smart_journal_bwrite(FAR struct smart_struct_s *dev, uint16_t physsector)
 {
 	int ret = OK;
+	int result = OK;
 	journal_log_t log;
-	struct smart_sect_header_s *header;
 	uint32_t address = 0;
-	size_t mtd_size = sizeof(FAR struct smart_sect_header_s);
+	uint16_t psector;
 
 	/* Initialize Journal log */
-	header = (FAR struct smart_sect_header_s *)dev->rwbuffer;
-	log.mtd_header = *header;
+	memcpy(&log.mtd_header, dev->rwbuffer, sizeof(FAR struct smart_sect_header_s));
 
 	/* Get Current Journal Address */
 	address = smart_journal_get_writeaddress(dev);
 
-	/* Block write follows this step, checkin -> write only data -> checkout */
 	ret = smart_journal_checkin(dev, &log, address, physsector, SMART_JOURNAL_TYPE_COMMIT);
 	if (ret != OK) {
-		/* If checkin failed, change state journal to checkout */
 		return ret;
 	}
 
-	/* Write given data in rwbuffer from users, except mtd header */
-	ret = MTD_WRITE(dev->mtd, physsector * dev->sectorsize + mtd_size, dev->sectorsize - mtd_size, (FAR uint8_t *)&dev->rwbuffer[mtd_size]);
-	if (ret != dev->sectorsize - mtd_size) {
-		fdbg("write data failed ret : %d\n", ret);
-		goto error_with_commit;
-	}
-
-	ret = smart_journal_checkout_process(dev, &log, address);
+	/* If block write transaction failed, we should release last sector to undo current transaction, and then checkout journal */
+	ret = smart_journal_process_transaction(dev, &log);
 	if (ret != OK) {
-		/* If checkout failed, we release data & change state journal to checkout to remove all data */
-
-		/* TODO Should we make journal for journal to handle This?? */
-		goto error_with_commit;
+		result = ret;
 	}
-
+	
+	if (result != OK) {
```

Comment:
This if statement can be merged with the if statement right above as follows:
if (ret != OK) {
  result = ret;
  psector = ...
  ret = smart_journal_release_sector();
  ....
}

Reply:
Yes right. I fixed it as you suggested, thanks.